lightning/impl/sag_fast.pyx

```diff
@@ -12,13 +13,49 @@ cimport numpy as np
 
 ctypedef np.int64_t LONG
 
-from libc.math cimport sqrt
+from libc.math cimport sqrt, abs, copysign
 
 from lightning.impl.randomkit.random_fast cimport RandomState
 from lightning.impl.dataset_fast cimport RowDataset
 from lightning.impl.sgd_fast cimport LossFunction
 
 
+cdef class Penalty:
+
+    cdef void projection(self, double* x, int* indices, int n_nz):
+        raise NotImplementedError()
+
+    cdef double regularization(self, double* x, int* indices, int n_nz):
+        raise NotImplementedError()
+
+
+cdef class L1Prox(Penalty):
+
+    cdef double l1
+
+    def __init__(self, double l1=0.):
+        self.l1 = l1
+
+    cdef void projection(self, double* x, int* indices, int n_nz):
+
+        cdef int j, jj
+
+        for jj in xrange(n_nz):
+            j = indices[jj]
+            x[j] = copysign(1., x[j]) * abs(x[j] - self.l1) * \
+                    (abs(x[j]) > self.l1)
+
```

Comment:
I am not sure this is doing what is expected since you are multiplying a float by a bool, and it is not guaranteed to behave as it does in Python.

Note that the soft thresholding can also be computed succintly as ```fmax(x[j] - l1, 0) - fmax(-x[j] - l1, 0)```.